# Introduction
* In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
* We sampled 50,000 data points from the full dataset for simple environment.
* The aim of this project is to clean the data and analyze the used car listings. 
---------------------------------------------------------

**Dictionary**
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Read file
autos = pd.read_csv("autos.csv", encoding = "Windows-1252")

In [3]:
# Inspect the data
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [32]:
# Inspect data
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 36998 entries, 0 to 49999
Data columns (total 20 columns):
datecrawled           36998 non-null object
name                  36998 non-null object
seller                36998 non-null object
offertype             36998 non-null object
price                 36998 non-null int64
abtest                36998 non-null object
vehicletype           32911 non-null object
registration_year     36998 non-null int64
gearbox               35010 non-null object
powerps               36998 non-null int64
model                 34957 non-null object
odometer_km           36998 non-null int64
registration_month    36998 non-null int64
fueltype              33413 non-null object
brand                 36998 non-null object
unrepaired_damage     28894 non-null object
ad_created            36998 non-null object
nrofpictures          36998 non-null int64
postalcode            36998 non-null int64
lastseen              36998 non-null object
dtypes: int64(7), 

* From the data, we have 20 features with 50,000 observations.
* The features are numeric and string.
* Column names use *Camel* style rather than *snakecase*.
* There are missing values in some features e.g., model, notRepairedDamage

In [5]:
# Explore column index
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# Rename variables with duplicate first element names
autos.rename(columns={'yearOfRegistration' : 'registration_year','monthOfRegistration' : 'registration_month', 'notRepairedDamage' : 'unrepaired_damage', 'dateCreated' : 'ad_created'}, inplace=True)

In [7]:
# The rest is to convert as lowercase (snakecase)
new_cols = []
for c in autos.columns:
    low = c.lower()
    new_cols.append(low)

# Value assignment
autos.columns = new_cols

# Check the new column names
autos.columns

Index(['datecrawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nrofpictures', 'postalcode',
       'lastseen'],
      dtype='object')

In [8]:
# Inspect the data again
autos.describe(include='all')

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

We can see that there are
* Columns with single value: `seller`, `offertype`
* Columns needed for deep dive: `registration_year`, `model`, `gearbox`, `vehicletype`, `fueltype`, `unrepaired_damage`
* Columns to convert to numeric: `price`, `odometer`

In [9]:
# Single value
print(autos["seller"].value_counts())
print('\n')
print(autos["offertype"].value_counts())

privat        49999
gewerblich        1
Name: seller, dtype: int64


Angebot    49999
Gesuch         1
Name: offertype, dtype: int64


In [10]:
# Wrong type
print(autos["price"].value_counts())
print('\n')
print(autos["odometer"].value_counts())

$0          1421
$500         781
$1,500       734
$2,500       643
$1,000       639
$1,200       639
$600         531
$800         498
$3,500       498
$2,000       460
$999         434
$750         433
$900         420
$650         419
$850         410
$700         395
$4,500       394
$300         384
$2,200       382
$950         379
$1,100       376
$1,300       371
$3,000       365
$550         356
$1,800       355
$5,500       340
$1,250       335
$350         335
$1,600       327
$1,999       322
            ... 
$1,797         1
$2,460         1
$27,299        1
$55,500        1
$2,695         1
$6,460         1
$16,995        1
$744           1
$566           1
$20,123        1
$3,260         1
$12,986        1
$29,550        1
$25,450        1
$15,298        1
$7,249         1
$1,955         1
$11,090        1
$29,250        1
$9,195         1
$15,777        1
$41,850        1
$4,239         1
$7,298         1
$220,000       1
$13,911        1
$4,220         1
$20,300       

In [11]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].str.strip().astype(int) # we can do this in one line.
autos["price"]

0         5000
1         8500
2         8990
3         4350
4         1350
5         7900
6          300
7         1990
8          250
9          590
10         999
11         350
12        5299
13        1350
14        3999
15       18900
16         350
17        5500
18         300
19        4150
20        3500
21       41500
22       25450
23        7999
24       48500
25          90
26         777
27           0
28        5250
29        4999
         ...  
49970    15800
49971      950
49972     3300
49973     6000
49974        0
49975     9700
49976     5900
49977     5500
49978      900
49979    11000
49980      400
49981     2000
49982     1950
49983      600
49984        0
49985     1000
49986    15900
49987    21990
49988     9550
49989      150
49990    17500
49991      500
49992     4800
49993     1650
49994     5000
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [12]:
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].str.strip().astype(int)
autos["odometer"]

0        150000
1        150000
2         70000
3         70000
4        150000
5        150000
6        150000
7        150000
8        150000
9        150000
10       150000
11       150000
12        50000
13       150000
14       150000
15        80000
16       150000
17       150000
18       150000
19       150000
20       150000
21       150000
22        10000
23       150000
24        30000
25       150000
26       125000
27       150000
28       150000
29       150000
          ...  
49970     60000
49971    150000
49972    150000
49973    150000
49974    150000
49975    100000
49976    150000
49977    150000
49978    150000
49979     70000
49980    125000
49981    150000
49982     90000
49983    150000
49984    150000
49985    150000
49986    125000
49987     50000
49988    150000
49989    150000
49990     30000
49991    150000
49992    125000
49993    150000
49994    150000
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 

In [13]:
# Rename 'odometer'
autos.rename({'odometer' : 'odometer_km'}, inplace=True, axis=1)

In [14]:
# Explore 'price'
print(autos["price"].describe())
print('\n')
print(autos["price"].unique().shape)
print('\n')
print(autos["price"].value_counts().sort_index(ascending=False).head(10))

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


(2357,)


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64


There are outliers in `price`

In [15]:
print(autos["price"].mean())
print(autos["price"].median())

9840.04376
2950.0


* The price data is skewed where median is lower than mean. Hence, we will use median and IQR to identify outliers.

In [16]:
# Remove outliers (median+1.5IQR, median-1.5IQR)

# find IQR
Q1 = autos["price"].quantile(0.25)
Q3 = autos["price"].quantile(0.75)
IQR = Q3-Q1

# Upper bound
upper = autos["price"].median() + (1.5 * IQR)
print(upper)

# Lower bound
lower = autos["price"].mean() - (1.5 * IQR)
print(lower)

# Removing outliers
autos = autos[autos["price"] <= upper] 

# Alternative -> df[df["col"].between(x,y)]

12100.0
690.0437600000005


In [17]:
# Check the price column again
print(autos["price"].describe())

count    43626.000000
mean      3363.079265
std       3026.563874
min          0.000000
25%        990.000000
50%       2350.000000
75%       4999.000000
max      12100.000000
Name: price, dtype: float64


In [18]:
# Explore 'odometer'
print(autos["odometer_km"].describe())
print('\n')
print(autos["odometer_km"].unique().shape)
print('\n')
print(autos["odometer_km"].value_counts().sort_index(ascending=False).head(10))

count     43626.000000
mean     131261.518361
std       35539.379615
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


(13,)


150000    30770
125000     4489
100000     1740
90000      1356
80000      1034
70000       816
60000       742
50000       576
40000       407
30000       392
Name: odometer_km, dtype: int64


In [19]:
print(autos["odometer_km"].mean())
print(autos["odometer_km"].median())

131261.51836061064
150000.0


In [20]:
# Remove outliers (median+1.5IQR, median-1.5IQR)

# Find IQR
Q1 = autos["odometer_km"].quantile(0.25)
Q3 = autos["odometer_km"].quantile(0.75)
IQR = Q3-Q1

# Upper bound
upper = autos["odometer_km"].median() + (1.5 * IQR)
print(upper)

# Lower bound
lower = autos["odometer_km"].mean() - (1.5 * IQR)
print(lower)

# Removing outliers
autos = autos[autos["odometer_km"].between(lower,upper)] 

187500.0
93761.51836061064


In [21]:
# Inspect 'odometer_km'
# Check the price column again
print(autos["odometer_km"].describe())

count     36999.000000
mean     144615.395011
std       12822.284121
min      100000.000000
25%      150000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


# Dealing with date columns
There are 5 columns that should represent date values. The attributes are constructed by either a crawler or the website itself. They are as follows;
* `date_crawled`: added by the crawler
* `last_seen`: added by the crawler
* `ad_created`: from the website
* `registration_month`: from the website
* `registration_year`: from the website

In [22]:
autos[['datecrawled','ad_created','lastseen']][0:5]

datecrawled           ad_created             lastseen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
5  2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21
6  2016-03-20 17:55:21  2016-03-20 00:00:00  2016-03-23 02:48:59

In [23]:
print(autos['datecrawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False))
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False))
print(autos['lastseen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False))

2016-04-07    0.001081
2016-04-06    0.003568
2016-04-05    0.013487
2016-04-04    0.036893
2016-04-03    0.037569
2016-04-02    0.034785
2016-04-01    0.033055
2016-03-31    0.031785
2016-03-30    0.033893
2016-03-29    0.033623
2016-03-28    0.034866
2016-03-27    0.030731
2016-03-26    0.032082
2016-03-25    0.032623
2016-03-24    0.028460
2016-03-23    0.032974
2016-03-22    0.032541
2016-03-21    0.037542
2016-03-20    0.037623
2016-03-19    0.033055
2016-03-18    0.013352
2016-03-17    0.031758
2016-03-16    0.030190
2016-03-15    0.033974
2016-03-14    0.037515
2016-03-13    0.015000
2016-03-12    0.037163
2016-03-11    0.032352
2016-03-10    0.032514
2016-03-09    0.033974
2016-03-08    0.034569
2016-03-07    0.036677
2016-03-06    0.014000
2016-03-05    0.024730
Name: datecrawled, dtype: float64
2016-04-07    0.000973
2016-04-06    0.003595
2016-04-05    0.012000
2016-04-04    0.037542
2016-04-03    0.037866
2016-04-02    0.034352
2016-04-01    0.033244
2016-03-31    0.031622


* From the above, popular days for crawling are found in several months e.g., 2016-04-03 (3.8%), 2016-03-21 (3.7%).
* The most popular day for the ad creation was 2016-04-03, with 3.8%.
* The most popular day for the last seen column was 2016-04-06, with 20.21%.

In [24]:
print(autos["registration_year"].describe())
print('\n')
print(autos["registration_year"].head(10))
print('\n')
print(autos["registration_year"].value_counts().sort_index())

count    36999.000000
mean      2002.422714
std         36.977419
min       1910.000000
25%       1998.000000
50%       2002.000000
75%       2006.000000
max       9000.000000
Name: registration_year, dtype: float64


0     2004
1     1997
4     2003
5     2006
6     1995
7     1998
8     2000
9     1997
10    2017
11    2000
Name: registration_year, dtype: int64


1910       2
1937       1
1950       1
1953       1
1954       1
1956       2
1958       1
1959       1
1960      14
1961       2
1962       1
1963       2
1964       1
1965       6
1966       5
1967       4
1968       6
1969       2
1970      15
1971       4
1972      13
1973      15
1974       6
1975       8
1976      13
1977      11
1978      21
1979      19
1980      51
1981      13
        ... 
1991     304
1992     330
1993     402
1994     585
1995    1172
1996    1328
1997    1898
1998    2270
1999    2792
2000    3037
2001    2479
2002    2313
2003    2461
2004    2419
2005    2546
2006    2132
2007    1610
2008    

We can see that there is an outlier, year 90000, which we are going to remove.

In [25]:
autos = autos[autos["registration_year"].between(1900,2020)] 
autos["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000054
1937    0.000027
1950    0.000027
1953    0.000027
1954    0.000027
1956    0.000054
1958    0.000027
1959    0.000027
1960    0.000378
1961    0.000054
1962    0.000027
1963    0.000054
1964    0.000027
1965    0.000162
1966    0.000135
1967    0.000108
1968    0.000162
1969    0.000054
1970    0.000405
1971    0.000108
1972    0.000351
1973    0.000405
1974    0.000162
1975    0.000216
1976    0.000351
1977    0.000297
1978    0.000568
1979    0.000514
1980    0.001378
1981    0.000351
          ...   
1990    0.008298
1991    0.008217
1992    0.008919
1993    0.010865
1994    0.015812
1995    0.031677
1996    0.035894
1997    0.051300
1998    0.061355
1999    0.075464
2000    0.082086
2001    0.067004
2002    0.062517
2003    0.066517
2004    0.065382
2005    0.068815
2006    0.057625
2007    0.043516
2008    0.034705
2009    0.024704
2010    0.012703
2011    0.008244
2012    0.003270
2013    0.000595
2014    0.000162
2015    0.000135
2016    0.030083
2017    0.0327

The top registration year is 2000, covering 8% of the total care registration between 1990 - 2020.

In [31]:
# Vehicles reistered in the past 30 years.
print((autos['registration_year']
 .value_counts(normalize=True)
 .sort_index(ascending=False)
 .head(30)
 .sum()
))

# Vehicles reistered in the past 10 years.
print((autos['registration_year']
 .value_counts(normalize=True)
 .sort_index(ascending=False)
 .head(10)
 .sum()
))

# Vehicles reistered in the past 5 years.
print((autos['registration_year']
 .value_counts(normalize=True)
 .sort_index(ascending=False)
 .head(5)
 .sum()
))

0.9794042921238987
0.09873506676036542
0.07376074382399048


98% of vehicles were registerd in the past 30 years.

In [26]:
# To calculate avg price of top common 20 brands

# Empty dictionary
avg_price_brands = {}
    
# Top 20 common brands
topbrands = autos['brand'].value_counts().index[:20]

for b in autos['brand'].value_counts().index[:20]:
    rows = autos[autos["brand"] == b]
    avg = rows['price'].mean()
    avg_price_brands[b] = avg #add value to the dictionary
    
avg_price_brands

{'audi': 4261.269666666667,
 'bmw': 4396.504785894206,
 'citroen': 2565.670542635659,
 'fiat': 1760.088888888889,
 'ford': 2199.1532075471696,
 'honda': 2472.0066445182724,
 'hyundai': 3027.871527777778,
 'mazda': 2537.935810810811,
 'mercedes_benz': 4095.9371479395195,
 'mitsubishi': 2083.343653250774,
 'nissan': 2601.1129943502824,
 'opel': 2038.8087167070219,
 'peugeot': 2218.515437392796,
 'renault': 1662.91473259334,
 'seat': 2609.294200848656,
 'skoda': 4032.80206185567,
 'smart': 2296.664935064935,
 'toyota': 3699.846153846154,
 'volkswagen': 3101.913506400287,
 'volvo': 2988.425587467363}

In [27]:
# Sorting the dictionary by values
sorted(avg_price_brands.items(), key = lambda x: x[1], reverse = True)
# More info here: https://realpython.com/python-sort/

[('bmw', 4396.504785894206),
 ('audi', 4261.269666666667),
 ('mercedes_benz', 4095.9371479395195),
 ('skoda', 4032.80206185567),
 ('toyota', 3699.846153846154),
 ('volkswagen', 3101.913506400287),
 ('hyundai', 3027.871527777778),
 ('volvo', 2988.425587467363),
 ('seat', 2609.294200848656),
 ('nissan', 2601.1129943502824),
 ('citroen', 2565.670542635659),
 ('mazda', 2537.935810810811),
 ('honda', 2472.0066445182724),
 ('smart', 2296.664935064935),
 ('peugeot', 2218.515437392796),
 ('ford', 2199.1532075471696),
 ('mitsubishi', 2083.343653250774),
 ('opel', 2038.8087167070219),
 ('fiat', 1760.088888888889),
 ('renault', 1662.91473259334)]

Among the top common 20 brands, BMW has the highest average price, $4,397, followed by Audi and Mercedes Benz. The bottom average price are Renault, Fiat, and Opel, respectively. This potentially signifies either brand values from customer demand perspective or brand positioning in the market.

In [28]:
# Empty dictionary
avg_mileage_brands = {} 
    
# Top 20 common brands
topbrands = autos['brand'].value_counts().index[:20]

for b in topbrands:
    rows = autos[autos['brand'] == b]
    avg_mileage = rows['odometer_km'].mean()
    avg_mileage_brands[b] = avg_mileage #add value to the dictionary
    
avg_mileage_brands

{'audi': 147566.66666666666,
 'bmw': 146945.84382871535,
 'citroen': 140891.47286821707,
 'fiat': 140211.64021164022,
 'ford': 143641.50943396226,
 'honda': 143770.76411960134,
 'hyundai': 139236.11111111112,
 'mazda': 143918.9189189189,
 'mercedes_benz': 146709.16098428698,
 'mitsubishi': 143808.04953560373,
 'nissan': 141713.74764595102,
 'opel': 144128.32929782083,
 'peugeot': 143267.58147512865,
 'renault': 143440.9687184662,
 'seat': 142892.5035360679,
 'skoda': 141185.56701030929,
 'smart': 134090.9090909091,
 'toyota': 140209.7902097902,
 'volkswagen': 145615.50424691947,
 'volvo': 148172.32375979112}

In [29]:
# Convert top_brands_mean_price dictionary into a Series
pb_series = pd.Series(avg_price_brands)
pb_series = pb_series.round(2)

# Convert top_brands_mean_mileage dictionary into a Series
mb_series = pd.Series(avg_mileage_brands)
mb_series = mb_series.round(2)

# Create new DataFrame to combine two series objects
comparison = pd.DataFrame(pb_series, columns=['mean_price'])
comparison['mean_mileage'] = mb_series   

# Display tb_mean_price_mileage sorted by price.
comparison.sort_values('mean_price', ascending=False)

mean_price  mean_mileage
bmw               4396.50     146945.84
audi              4261.27     147566.67
mercedes_benz     4095.94     146709.16
skoda             4032.80     141185.57
toyota            3699.85     140209.79
volkswagen        3101.91     145615.50
hyundai           3027.87     139236.11
volvo             2988.43     148172.32
seat              2609.29     142892.50
nissan            2601.11     141713.75
citroen           2565.67     140891.47
mazda             2537.94     143918.92
honda             2472.01     143770.76
smart             2296.66     134090.91
peugeot           2218.52     143267.58
ford              2199.15     143641.51
mitsubishi        2083.34     143808.05
opel              2038.81     144128.33
fiat              1760.09     140211.64
renault           1662.91     143440.97

Among the top common 20 brands, we observe that the range between the higest average price and the lowest one (\$1,663 - $4,396), is bigger than the range of the average mileage (143,441 km - 146,946 km).

# Conclusion
